### PR TITLE
[PyTorch] Reconstruct output args to fit pytorch and other bug fixes

### PR DESCRIPTION
**Pull request by Atrisan**
## Description
Reorders result tuples to fit the original pytorch output.
Generate copy kernels where the strides and tensor layout are regenerated instead of being ignored.
Fix the 2 broadcast index to the correct dimension.
Skip dynamic size parsing with torch compile.

## Related Issues

- Closes #

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [x] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [x] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [ ]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [ ] New / modified types have serialization / deserialization support
  *  [ ] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [ ] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [x] New unit tests have been added for the changes.
- [x] Existing unit tests pass locally.
- [x] Edge cases and error paths are covered.

### Integration Tests
- [x] New integration tests have been added (or existing ones updated).
- [x] Integration tests pass locally.
- [x] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [x] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [x] Test coverage has not decreased as a result of this change.
- [x] Coverage report reviewed and acceptable.
- [x] Critical / new code paths are covered.

## How Has This Been Tested?
Describe the tests you ran to verify your changes and provide instructions to reproduce.
Include relevant details for your test configuration (OS, Python/LLVM/Docc version, hardware/target).

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: pytorch/docc/pytorch/graph_parser/convolution.py [hunk 1]
```diff
@@ -60,14 +60,31 @@ def parse(
...
         result_tensor = self.get_tensor_type(node, container_info, result_container)
         debug_info: DebugInfo = self.get_debug_info(node)
+        # DOCC's convolution expansion always writes a contiguous (NCHW) output.
+        # When torch.compile selects a non-contiguous memory format for the
+        # convolution (e.g. channels_last), the fx metadata expresses the result
+        # -- and every downstream view/permute derived from it -- relative to that
+        # layout. Emit the convolution into a contiguous buffer and copy it into
+        # the requested (strided) layout so the physical data matches what the
+        # consumers assume; otherwise the layout mismatch silently transposes the
+        # data (observed as a wrong LayerNorm result in SegFormer).
```
Comment (WashimNeupane) [inline pytorch/docc/pytorch/graph_parser/convolution.py:70]: creating a intermediate tenosr will cause performance/memory issues down the line. the fix is ok for now for correctness test, but a better alternative would be to change the conv library node to accept non-contiguous output i.e. move away from expanding to blas and into cuDNN, MIOpen, ACL, etc.

Comment (Atrisan) [reply pytorch/docc/pytorch/graph_parser/convolution.py:70]: Not necessarily, thanks to the map fusion many of these interim results and manifestations can actually be fused into their predecessor or successor, as long as they at least one of them expands into a loopnest instead of another libnode. A different lib does not really help here.

File: sdfg/include/sdfg/data_flow/library_nodes/math/tensor/broadcast_node.h [hunk 2]
```diff
@@ -14,6 +14,8 @@ class BroadcastNode : public TensorNode {
 private:
     std::vector<symbolic::Expression> input_shape_;
     std::vector<symbolic::Expression> output_shape_;
+    // true: leading-alignment (linalg.broadcast); false: trailing-alignment (NumPy/PyTorch).
+    bool padded_;
```
Comment (WashimNeupane) [inline sdfg/include/sdfg/data_flow/library_nodes/math/tensor/broadcast_node.h:18]: maybe something like is_left_aligned would be more appropriate/intuitive than padded_ here.

File: pytorch/docc/pytorch/graph_parser/convolution.py (hunk 1 above)
Comment (WashimNeupane) [reply pytorch/docc/pytorch/graph_parser/convolution.py:70]: makes sense.

**Review by WashimNeupane — DISMISSED** [review]

File: .github/workflows/unit_tests_linux.yml [hunk 3]
```diff
@@ -5,11 +5,29 @@ on:
...
-    types: [opened, reopened, synchronize, ready_for_review]
+    types: [opened, reopened, synchronize]
   schedule:
     - cron: "0 4 * * *"
 
 jobs:
+  setup:
+    runs-on: ubuntu-latest
+    outputs:
+      python-versions: ${{ steps.matrix.outputs.python-versions }}
```
Comment (lukastruemper) [inline .github/workflows/unit_tests_linux.yml]: I feel like this is a bit overcomplicating things. Just split the workloads. Maybe as a separate PR?

**Review by lukastruemper — CHANGES_REQUESTED** [review]

File: .github/workflows/unit_tests_linux.yml (hunk 3 above)
Comment (Atrisan) [reply .github/workflows/unit_tests_linux.yml]: A separate PR would be reasonable.
The idea behind this was to keep the tests identical and a oid replication. Thereby, encuraging that the tests cannot diverge in future changes as well.

File: pytorch/tests/__init__.py [hunk 4]
```diff
@@ -72,6 +72,14 @@ def check(
...
         model,
         backend="docc",
+        # Specialize on concrete input shapes. Without this, Dynamo's automatic
+        # dynamic-shape behavior turns the batch dimension into a symbolic SymInt
+        # on the second compile of the same model code (e.g. the batch_size=1 and
+        # batch_size=4 parametrizations of the same test). That makes the traced
+        # example_input shapes identical (s0, ...) across batch sizes, so the
+        # compile cache key / stable_id collide and both batches reuse the same
+        # SDFG and .so. Forcing static shapes gives each batch size its own graph.
+        dynamic=False,
```
Comment (lukastruemper) [inline pytorch/tests/__init__.py:82]: I thought this is what we want, dynamic sizes?

**Review by lukastruemper — APPROVED** [review]
LGTM, but thought we want dynamic sizes

File: pytorch/tests/__init__.py (hunk 4 above)
Comment (Atrisan) [reply pytorch/tests/__init__.py:82]: I agree, but we currently do not support them and if we calculate the same hash, pytorch automatically defaults to dynamic sizes, causing issues in testing.
Once we have support it can be reactivated, but this keeps the behaviour consistent.
